impeller/entity/entity_pass.cc

```diff
@@ -174,231 +178,249 @@ bool EntityPass::Render(ContentContext& renderer,
+  /// Setup subpass element.
+  ///
 
-  std::shared_ptr<CommandBuffer> command_buffer;
-  std::shared_ptr<RenderPass> pass;
-  uint32_t pass_count = 0;
+  else if (const auto& subpass_ptr =
+               std::get_if<std::unique_ptr<EntityPass>>(&element)) {
+    auto subpass = subpass_ptr->get();
 
-  auto end_pass = [&command_buffer, &pass, &context]() {
-    if (!pass->EncodeCommands(context->GetTransientsAllocator())) {
-      return false;
+    if (subpass->delegate_->CanElide()) {
+      return EntityPass::EntityResult::Skip();
     }
 
-    if (!command_buffer->SubmitCommands()) {
-      return false;
-    }
-
-    return true;
-  };
-
-  for (const auto& element : elements_) {
-    Entity element_entity;
-
-    // =========================================================================
-    // Setup entity element for rendering ======================================
-    // =========================================================================
-    if (const auto& entity = std::get_if<Entity>(&element)) {
-      element_entity = *entity;
-      if (!position.IsZero()) {
-        // If the pass image is going to be rendered with a non-zero position,
-        // apply the negative translation to entity copies before rendering them
-        // so that they'll end up rendering to the correct on-screen position.
-        element_entity.SetTransformation(
-            Matrix::MakeTranslation(Vector3(-position)) *
-            element_entity.GetTransformation());
+    if (subpass->delegate_->CanCollapseIntoParentPass() &&
+        !subpass->backdrop_filter_proc_.has_value()) {
+      // Directly render into the parent target and move on.
+      if (!subpass->OnRender(renderer, pass_context.GetRenderTarget(), position,
+                             position, stencil_depth_floor)) {
+        return EntityPass::EntityResult::Failure();
       }
+      return EntityPass::EntityResult::Skip();
     }
 
-    // =========================================================================
-    // Setup subpass element for rendering =====================================
-    // =========================================================================
-    else if (const auto& subpass_ptr =
-                 std::get_if<std::unique_ptr<EntityPass>>(&element)) {
-      auto subpass = subpass_ptr->get();
-
-      if (subpass->delegate_->CanElide()) {
-        continue;
-      }
+    std::optional<Rect> subpass_coverage;
 
-      if (subpass->delegate_->CanCollapseIntoParentPass()) {
-        // Directly render into the parent target and move on.
-        if (!subpass->RenderInternal(renderer, render_target, position,
-                                     pass_depth, stencil_depth_floor)) {
-          return false;
-        }
-        continue;
-      }
+    if (subpass->backdrop_filter_proc_.has_value()) {
+      subpass_coverage = Rect(
+          position, Size(pass_context.GetRenderTarget().GetRenderTargetSize()));
```

Comment:
It won't really happen with a blur, but it is possible for the filtered output to be smaller than the saveLayer rect. Consider the erode filter, for instance. With a large erode radius it could drop the total size of the filtered backdrop down to a very small size.

Although you don't seem to be computing the filtered size here, just the size of the backdrop...?

Reply:
I did some more comprehensive testing in Skia fiddle and confirmed that coverage of the filter is important. Right now there are 2 problems with my coverage approach here:
1. As you mentioned, the post-filter coverage is not getting incorporated. This means that if you use a blur filter on a small layer, the expanded sides of the blur will get cut off (which is wrong).
2. _Only_ the backdrop texture is getting incorporated when a filter is present, which means entities added to the layer  with the filter may get cut off, which is also wrong.

Reply:
In #1, by "small layer", do you mean the saveLayer bounds? Presumably if the parent layer is small, the blur would be clipped to it anyway.

For #2, I'm not sure what you mean by "entities added to the layer"? Do you mean things you draw in the sub-pass, or the content of the saveLayer? I think those would normally (in Skia?) be cut off if the saveLayer bounds were provided without a backdrop filter, but not if the backdrop filter requires a larger layer.

My impression is that the bounds are:

Union(backdrop(parent layer).output_bounds, saveLayer bounds).IntersectWith(clip and parent layer bounds).

Reply:
> Presumably if the parent layer is small, the blur would be clipped to it anyway

This is what I thought too, but the blur seems to be getting properly incorporated into the layer's coverage (which in turn impacts the parent layer's size): https://fiddle.skia.org/c/7cb23795f3cfc78a10e50c3503379d38
In the above Skia fiddle, I expected to see the blur to get cut off on the bottom and right side of the middle circle (#2), But to my surprise, the parent layer's coverage seems to expand appropriately and nothing is getting cut off.

There's a dependency loop. Assuming SaveLayer B is a child of SaveLayer A, and SaveLayer B has a backdrop filter:
1. The SaveLayer A's coverage depends on the coverage of SaveLayer B.
2. The SaveLayer B's coverage depends on the backdrop filter coverage.
3. The backdrop filter coverage depends on the parent pass texture size.
4. The parent pass texture size is the SaveLayer A's coverage.

This can be solved by extending the filter coverage API to allow the caller to supply a list of prospective coverage rectangles for the inputs (as opposed to resolving the `FilterInput` coverage) via the `Contents::GetCoverage` implementation. Then, `EntityPass` coverage routine can incorporate subpass filter coverage without having to allocate textures. Another alternative would be to add mock/coverage-only filter inputs.

Reply:
If you specify bounds for A's saveLayer then those get used as a clip, but yeah. B's saveLayer is a rendering operation within A so its bounds affect the default bounds for A. Also, if you applied a blur filter to B (not a backdrop, but a blur of the layer), those blur bounds would also be accumulated into A's default bounds.

Reply:
Alright, I changed the coverage computation to union with the backdrop filter's coverage before creating the child pass texture.

However, this union currently happens outside the actual coverage routine, which means the backdrop filter coverage still doesn't properly propagate up to parent passes. Fixing this is a bit more involved, so I wrote an issue with a plan for how to address it in a future patch: https://github.com/flutter/flutter/issues/105810